pkg/cmd/codespace/jupyter.go

```diff
@@ -45,13 +45,7 @@ func (a *App) Jupyter(ctx context.Context, codespaceName string) (err error) {
 	defer safeClose(session, &err)
 
 	a.StartProgressIndicatorWithLabel("Starting JupyterLab on codespace")
-	client, err := connectToGRPCServer(ctx, session, codespace.Connection.SessionToken)
-	if err != nil {
-		return fmt.Errorf("failed to connect to internal server: %w", err)
-	}
-	defer safeClose(client, &err)
-
-	serverPort, serverUrl, err := startJupyterServer(ctx, client)
+	serverPort, serverUrl, err := rpc.StartJupyterServer(ctx, session)
```

Comment:
Defining functions on the package level (`rpc.StartJupyterServer(ctx, session)`) instead of on the type level (`session.StartJupyterServer(ctx)`) makes it harder to pass in a mocked session implementation in test mode.

What do you think about an approach where there is a session type that implements the old `liveshareSession` interface, but invokes specific methods over RPC instead of to Liveshare? Example use:

```go
session := startSession(ctx, codespace)
session.StartSSHServer(ctx) // goes to liveshare
session.StartJupyterServer(ctx) // goes to RPC
```

That way, any codespace command that currently interacts with a session and invokes methods on it doesn't have to be aware that the underlying implementation for those methods has completely changed to be backed by RPC. The gradual adoption of RPC would be an implementation detail.

Reply:
I was going to suggest something similar, but putting it in `grpc.Client` instead. We already pass in the `session` to the `connectToGRPCServer` "constructor" and so we shouldn't need to keep passing in the session for each RPC call we make. In that case, `grpc.Client` can do all the redirects to `session` instead of the package level function.

For example:

```go
func RebuildContainer(ctx context.Context, session *liveshare.Session, full bool) error {
	return session.RebuildContainer(ctx, full)
}
```

becomes

```go
func (c *grpc.Client) RebuildContainer(ctx context.Context, full bool) error {
	return c.session.RebuildContainer(ctx, full)
}
```

Since we are migrating away from liveshare, I do agree with @dmgardiner25 that we should also migrate off the `session` name as `session` is a liveshare concept and an implementation detail.

Reply:
Updated to use `invoker` instead of `grpc` and should hopefully be more in line with what you were thinking! Let me know if you have any other thoughts 🙂